src/forwardBeamPipeBrazil.cpp

```diff
@@ -0,0 +1,570 @@
+
+        //------------------------------------------------------------------------------------
+    // pipe connecting taper to B2PF magnet, just past ZDC
+    //------------------------------------------------------------------------------------
+
+    //numbers here are not really correct for the full taper, just for the opening
+
+    length[7]            = 3000.0; // from VPC drawings
+    innerXRadius[7]      = 35.0; //
+    innerYRadius[7]      = 0.0; //NOT USED
+    outerXRadius[7]      = innerXRadius[7] + pipeThickness; //
+    outerYRadius[7]      = innerYRadius[7] + pipeThickness; //NOT USED
+
+    xCenter[7]           = -1*(tmp_endpoint_x+((0.5*length[7])*TMath::Sin(-globRotationAngle)));//-972.36849;
+    yCenter[7]           = 0.0;
+    zCenter[7]           = tmp_endpoint_z+((0.5*length[7])*TMath::Cos(-globRotationAngle));//29055.1545;
+    rotationAngle[7]     = globRotationAngle;
+
+
+        //------------------------------------------
+        //begin building main volumes here
+        //------------------------------------------
+
+        int pieceIdx = 0; //Larger, rectangular pipe transporting proton and neutral envelopes (neutral exit window and transfer to smaller proton line at the end)
+
+        Box inner_subtract_outer(dd4hep::mm*outerXRadius[5], dd4hep::mm*outerYRadius[5], dd4hep::mm*length[5]/2);
+        Box inner_subtract_inner(dd4hep::mm*innerXRadius[5], dd4hep::mm*innerYRadius[5], dd4hep::mm*(length[5]+5.0)/2);
+
+        Box RP_subtract_outer(dd4hep::mm*outerXRadius[1], dd4hep::mm*outerYRadius[1], dd4hep::mm*(length[2]+5.0)/2);
+        Box RP_subtract_inner(dd4hep::mm*innerXRadius[1], dd4hep::mm*innerYRadius[1], dd4hep::mm*(length[2]+10.0)/2);
+
+        Box pipeAfterB1APF_outer(dd4hep::mm*outerXRadius[pieceIdx], dd4hep::mm*outerYRadius[pieceIdx], dd4hep::mm*length[pieceIdx]/2);
+        Box pipeAfterB1APF_inner(dd4hep::mm*innerXRadius[pieceIdx], dd4hep::mm*innerYRadius[pieceIdx], dd4hep::mm*(length[pieceIdx])/2);
+        Box pipeAfterB1APF_firstEndCap(dd4hep::mm * outerXRadius[pieceIdx], dd4hep::mm * outerYRadius[pieceIdx], dd4hep::mm*5.0/2.0);
+        Tube neutral_exit_window_cutout(0.0, dd4hep::mm*windowRadius, 1.0); // 1.0cm thick
+        Box protonTransferWindow(dd4hep::mm * 155.0, dd4hep::mm*outerYRadius[1], dd4hep::mm*(5.0/2));
+
+        SubtractionSolid tmpAfterB1APF(pipeAfterB1APF_outer, pipeAfterB1APF_inner); //This gets rid of the inner portion of the pipe, but leaves the endcaps
+        SubtractionSolid tmpAfterFrontEndCap(tmpAfterB1APF, pipeAfterB1APF_firstEndCap, Position(0.0, 0.0, dd4hep::mm * (-length[pieceIdx])/2));
+        SubtractionSolid pipeAfterProtonTransferWindow(tmpAfterFrontEndCap, protonTransferWindow, Position(dd4hep::mm * (-120.0), 0.0, dd4hep::mm * (length[pieceIdx])/2 ));
+
+        SubtractionSolid pipeAfterB1APF(pipeAfterProtonTransferWindow, neutral_exit_window_cutout, Position(dd4hep::mm * 160.0 , 0.0, dd4hep::mm * 0.5*length[pieceIdx]));
```

Comment:
Every serial subtraction solid should be placed inside the outermost regular solid (here: `pipeAfterB1APF_outer`). Subtraction solids are expensive to determine track intersections with. Quadruple subtractions are more than quadruply so since some heuristics that optimize intersection calculations can't be used. By placing this inside its own volume, we ensure that this won't affect every step in the far forward direction.

Reply:
Done.